cli/stork-publisher-agent/runner.go

```diff
@@ -0,0 +1,281 @@
+	if err != nil {
+		r.logger.Error().Err(err).Msg("failed to get broker connections from Stork Registry")
+		return
+	}
+
+	// add or update desired connections
+	for brokerUrl, newAssetIdMap := range newBrokerMap {
+		_, exists := r.brokerMap[brokerUrl]
+		if exists {
+			// update connection
+			r.outgoingConnections[brokerUrl].UpdateAssets(newAssetIdMap)
+		} else {
+			// create connection
+			go r.RunOutgoingConnection(brokerUrl, newAssetIdMap, r.config.StorkAuth)
+		}
+		r.brokerMap[brokerUrl] = newAssetIdMap
+	}
+
+	// remove undesired connections
+	for url, _ := range r.brokerMap {
+		_, exists := newBrokerMap[url]
+		if !exists {
+			r.outgoingConnections[url].Remove()
+			delete(r.brokerMap, url)
+		}
+	}
+
+	r.logger.Info().Msg("Broker connection updater finished")
+}
+
+func (r *PublisherAgentRunner[T]) RunBrokerConnectionUpdater() {
+	r.UpdateBrokerConnections()
+	for range time.Tick(r.config.StorkRegistryRefreshInterval) {
+		r.UpdateBrokerConnections()
+	}
+}
+
+func (r *PublisherAgentRunner[T]) Run() {
+	processor := NewPriceUpdateProcessor[T](
+		r.signer,
+		r.config.ClockPeriod,
+		r.config.DeltaCheckPeriod,
+		r.config.ChangeThresholdProportion,
+		r.config.SignEveryUpdate,
+		r.priceUpdateCh,
+		r.signedPriceBatchCh,
+		r.logger,
+	)
+
+	// fan out the signed update to all subscriber websockets
+	go func(signedPriceBatchCh chan SignedPriceUpdateBatch[T]) {
+		for signedPriceUpdateBatch := range signedPriceBatchCh {
+			r.outgoingConnectionsLock.RLock()
+			for _, outgoingConnection := range r.outgoingConnections {
+				outgoingConnection.signedPriceUpdateBatchCh <- signedPriceUpdateBatch
+			}
+			r.outgoingConnectionsLock.RUnlock()
+		}
+	}(r.signedPriceBatchCh)
+
+	go r.RunBrokerConnectionUpdater()
+
+	if len(r.config.PullBasedWsUrl) > 0 {
+		go r.RunPullBasedIncomingConnection(
+			r.config.PullBasedWsUrl,
+			r.config.PullBasedAuth,
+			r.config.PullBasedWsSubscriptionRequest,
+			r.config.PullBasedWsReconnectDelay,
+		)
+	}
+
+	processor.Run()
+}
+
+func (r *PublisherAgentRunner[T]) RunPullBasedIncomingConnection(url string, auth AuthToken, subscriptionRequest string, reconnectDelay time.Duration) {
+	for {
+		r.logger.Info().Msgf("Connecting to pull-based WebSocket with url %s", url)
+
+		var headers http.Header
+		if len(auth) > 0 {
+			headers = http.Header{"Authorization": []string{"Basic " + string(auth)}}
+		}
+
+		incomingWsConn, _, err := websocket.DefaultDialer.Dial(url, headers)
+		if err != nil {
+			r.logger.Error().Err(err).Msgf("Failed to connect to pull-based WebSocket: %v", err)
+			break
+		}
+
+		_, messageBytes, err := incomingWsConn.ReadMessage()
+		if err != nil {
+			r.logger.Error().Err(err).Msgf("Failed to read connection message from pull-based WebSocket: %v", err)
+		}
+		r.logger.Info().Msgf("Received connection message: %s", messageBytes)
+
+		if len(subscriptionRequest) > 0 {
+			r.logger.Info().Msgf("Sending subscription request: %s", subscriptionRequest)
```

Comment:
Lets do a quick audit of all of these actually

Reply:
went through all the log statements and everything expected to Debug, so Info will only be expected failures like websockets closing or being intentionally removed and real problems will be warn or error.